### PR TITLE
SDK support for the new OTel log API

### LINKED ISSUE #2122: Support new user-facing log API in SDK
As #2094 is merged, SDK should also be updated to support the new API.

### PR DESCRIPTION
**Pull request by ThomsonTan**
Fixes #2122 

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

File: api/include/opentelemetry/logs/logger_type_traits.h [hunk 1]
```diff
@@ -48,9 +48,22 @@ template <>
...
+  inline static LogRecord *Set(LogRecord *log_record, ArgumentType &&arg) noexcept
+  {
+    log_record->SetEventId(arg.id_, nostd::string_view{arg.name_.get()});
+
+    return log_record;
+  }
+};
+
+template <>
+struct LogRecordSetterTrait<int64_t>
```
Comment (owent) [inline api/include/opentelemetry/logs/logger_type_traits.h:60]: Is it conflicted with Body?it can also be converted from `int64_t` implicitly.

Comment (ThomsonTan) [reply api/include/opentelemetry/logs/logger_type_traits.h:60]: Yes, thanks for the catch. Seems the template part of filling logging record requires each parameter to be in different type, and the `any` type on body make it conflict with other parameters. Do you have any suggestion for this problem?

Comment (owent) [reply api/include/opentelemetry/logs/logger_type_traits.h:60]: I have no idea for this problem by now, and I only avoided to specialize templates for these types before.
Maybe we can add a unit test to check the mistake.

Comment (ThomsonTan) [reply api/include/opentelemetry/logs/logger_type_traits.h]: I removed the passing of int64_t to SDK, instead, a temporary EventId struct will be constructed which should fix the issue.

For unittest, do you mean to cover all the variant types of `body` field?

File: api/include/opentelemetry/logs/event_id.h [hunk 2]
```diff
@@ -18,14 +18,16 @@ namespace logs
 class EventId
 {
 public:
-  EventId(int64_t id, nostd::string_view name) noexcept
+  EventId(int64_t id, nostd::string_view name) noexcept : id_{id}
```
Comment (lalitb) [inline api/include/opentelemetry/logs/event_id.h:21]: not required to initialize `id_` inside constructor body if it is initialized using initializer list.

**Review by lalitb — APPROVED** [review]
LGTM. Thanks

File: api/include/opentelemetry/logs/event_id.h (hunk 2 above)
Comment (ThomsonTan) [reply api/include/opentelemetry/logs/event_id.h:21]: Thanks for the catch.

File: exporters/otlp/include/opentelemetry/exporters/otlp/otlp_log_recordable.h [hunk 3]
```diff
@@ -67,6 +67,16 @@ class OtlpLogRecordable final : public opentelemetry::sdk::logs::Recordable
    */
   void SetBody(const opentelemetry::common::AttributeValue &message) noexcept override;
 
+  /**
+   * @brief Set the Event Id for this log.
+   * @param id the event Id to set
+   * @param name  the event name to set
+   */
+  void SetEventId(int64_t /* id */, nostd::string_view /* name */) noexcept override
```
Comment (owent) [inline exporters/otlp/include/opentelemetry/exporters/otlp/otlp_log_recordable.h:75]: Did I miss anything? I can find any specification about event id in OTLP, should we add it just like [semantic convention for event attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/semantic_conventions/events.md)?

File: api/include/opentelemetry/logs/logger_type_traits.h (hunk 1 above)
Comment (owent) [reply api/include/opentelemetry/logs/logger_type_traits.h]: Yes, use EmitLogRecord to cover all the variant types of body field.
Maybe we can add it in another PR later.

File: sdk/include/opentelemetry/sdk/logs/multi_recordable.h [hunk 4]
```diff
@@ -55,6 +55,14 @@ class MultiRecordable final : public Recordable
    */
   void SetBody(const opentelemetry::common::AttributeValue &message) noexcept override;
 
+  /**
+   * Set the Event Id
+   * @param id the event id to set
+   * @param name  the event name to set
+   */
+  // TODO: implement set event id
+  void SetEventId(int64_t /* id */, nostd::string_view /* name */) noexcept override {}
```
[the remark this answers is not in the document]
Comment (ThomsonTan) [reply sdk/include/opentelemetry/sdk/logs/multi_recordable.h]: Thanks. Yes, the `multi_recordable` method should just forward the call to each individual recordable. Updated with this.

File: api/include/opentelemetry/logs/logger_type_traits.h (hunk 1 above)
Comment (ThomsonTan) [reply api/include/opentelemetry/logs/logger_type_traits.h]: FYI, I added an integration test to cover if the EventId passed to the user-facing API is passed correctly to the SDK.

https://github.com/ThomsonTan/opentelemetry-cpp/blob/otel_cpp_log_sdk_impl/exporters/ostream/test/ostream_log_test.cc#L465

**Comment by ThomsonTan** [comment]
@owent please let me know if there is any comments I missed.

**Pull request merged**